### PR TITLE
feat(visualsign): Anchorage wallet render-support validator

### PR DESCRIPTION
**Pull request by prasanna-anchorage**
## What
A validator that checks a `SignablePayload` only uses fields the **Anchorage wallet** can render, mirroring the wallet's VSP field support matrix — so the parser can catch payloads that would surface as "unsupported fields" before they ship (VSP is WYSIWYS, so an unrenderable field is a correctness bug).

- `visualsign::SignablePayload::validate_anchorage_wallet_renderable()` — hard gate (errs listing offending paths).
- `anchorage_render_findings()` — every unrenderable field with a JSON-ish path (e.g. `Fields[4].Expanded.Fields[6]`), label, type, and reason. Flags `number`, legacy `text`/`address`/`amount`, `divider`, `unknown`, and `list_layout`-as-a-standalone-field; recurses through `preview_layout` propagating `ContainsUnsupportedNestedFields`.
- `examples/check_anchorage_render.rs` — `cargo run -p visualsign --features diagnostics --example check_anchorage_render -- <payload.json>`.
- `docs/specs/anchorage-wallet-render-rules.md` — the authoritative ruleset (source of truth: the Anchorage wallet's Visual Signing Protocol engine).

Draft — opening for visibility; a natural follow-up is wiring it into CI alongside `validate_charset`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/visualsign/examples/check_anchorage_render.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+use std::fs;
+
+use visualsign::SignablePayload;
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let path = std::env::args()
+        .nth(1)
+        .ok_or("usage: check_anchorage_render <payload.json>")?;
+    let json = fs::read_to_string(&path)?;
```
[the remark this answers is not in the document]
Comment (prasanna-anchorage) [reply src/visualsign/examples/check_anchorage_render.rs:15]: This compiles as-is: `?` converts the `&'static str` error into `Box<dyn Error>` via std's `From<&str>` impl (verified with `cargo build --example check_anchorage_render`). No change needed.

File: src/visualsign/src/anchorage_render.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+//! | `text` (v1)      | no                 | superseded by `text_v2`                 |
+//! | `address` (v1)   | no                 | superseded by `address_v2`              |
+//! | `amount` (v1)    | no                 | superseded by `amount_v2`               |
+//! | `divider`        | no                 | not in the wallet decoder               |
+//! | `unknown`        | no                 | explicit fallback/unsupported           |
+//!
+//! The wallet additionally renders `delta`, `highlight`, `rule`, and `accordion`
+//! containers, but the parser has no field variant that emits them, so they
+//! cannot occur in parser output. They are listed in
+//! [`ANCHORAGE_RENDERABLE_LEAF_TYPES`] for completeness / forward compatibility.
```
[the remark this answers is not in the document]
Comment (prasanna-anchorage) [reply src/visualsign/src/anchorage_render.rs]: Fixed in caaa15cb: reworded so delta/highlight/rule are described as leaf types and only accordion as a container, consistent with ANCHORAGE_RENDERABLE_LEAF_TYPES.

File: src/visualsign/src/anchorage_render.rs [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+            out.push(finding(
+                path,
+                field,
+                AnchorageUnsupportedReason::ListLayoutAsStandaloneField,
+            ));
+            false
+        }
+        // Every other variant is a leaf: renderable iff its type is in the set.
+        _ => {
+            if ANCHORAGE_RENDERABLE_LEAF_TYPES.contains(&field.field_type()) {
```
Comment (pepe-anchor) [inline src/visualsign/src/anchorage_render.rs]: should: `check_field`'s catch-all arm checks `field.field_type()`, a hardcoded per-variant literal in `lib.rs`, instead of the actual `"Type"` string `serialize_to_map()` puts on the wire. These are two independent match statements over the same enum that happen to agree today. Sibling PR #393 (open, same author, targets this PR's own documented `number` gap) changes only `serialize_to_map`'s `Number` arm to emit `amount_v2` on the wire while leaving `field_type()` returning `number`. Once #393 merges, every `Number` field will render fine in the wallet but this validator will flag it as unsupported forever, a permanent false positive baked in before the PR's own stated CI-wiring follow-up lands.

```suggestion
Derive the checked type string from `serialize_to_map()`'s `Type` entry (or a shared single source of truth) instead of `field_type()`, so the validator tracks whatever actually reaches the wallet on the wire.
```

File: src/visualsign/src/anchorage_render.rs [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+                }
+            }
+            if let Some(expanded) = &preview_layout.expanded {
+                for (index, child) in expanded.fields.iter().enumerate() {
+                    let ok = check_field(
+                        &child.signable_payload_field,
+                        &format!("{path}.Expanded.Fields[{index}]"),
+                        out,
+                    );
+                    all_ok = all_ok && ok;
```
Comment (pepe-anchor) [inline src/visualsign/src/anchorage_render.rs]: should: When `preview_layout.condensed` or `.expanded` is `None`, the `if let Some` branch is skipped: no descendants are checked and no finding is pushed. This is live today: `field_builders::create_preview_layout` always sets `condensed: None`, and multiple Ethereum visualizers (`erc20.rs`, `universal_router.rs`, `permit2.rs`, `erc1155.rs`) set `condensed`/`expanded` to `None`, e.g. the ERC-20 `Transfer` preview. `lib.rs` still has `skip_serializing_if = "Option::is_none"` on both fields on this PR's base, and open PR #403 documents that this exact shape drops the key from the JSON and throws `keyNotFound` on iOS, degrading the whole preview_layout to FallbackText in production (Sentry IOS-646/PRS-548). The validator calls this construct clean today even though it is a known, previously-shipped WYSIWYS decode failure; it becomes correct only once #403 merges.

```suggestion
PR #403 (https://github.com/anchorageoss/visualsign-parser/pull/403) fixes the serialization gap (always emitting Condensed/Expanded), but this validator should be fixed independently rather than relying on merge order: add a test that pins the None-handling behavior explicitly (e.g. assert a None condensed/expanded is not silently reported as clean, or check against the actual serialized wire output instead of the in-memory Option), so this code stays correct on its own regardless of whether #403 has landed.
```

File: src/visualsign/examples/check_anchorage_render.rs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+//! Report the fields an Anchorage wallet cannot render in a SignablePayload JSON.
+//!
+//! Usage:
+//!   cargo run -p visualsign --features diagnostics \
+//!     --example check_anchorage_render -- <payload.json>
```
Comment (pepe-anchor) [inline src/visualsign/examples/check_anchorage_render.rs:5]: nit: The example builds and runs fine without `--features diagnostics` for ordinary payloads, but a payload containing `"Type": "diagnostic"` (one of the four renderable leaf types) fails to deserialize at all without the feature: `unknown variant 'diagnostic'`. The documented flag is real, not decorative, but forgetting it surfaces a generic serde error instead of a hint.

```suggestion
Catch the deserialize error and print a hint to pass --features diagnostics, or default-enable the feature for this example.
```

File: src/visualsign/src/anchorage_render.rs (hunk 3 above)
Comment (prasanna-anchorage) [reply src/visualsign/src/anchorage_render.rs]: Confirmed — checked current `main`: PR #393 (merged as 52eda7f4) makes `Number` serialize as `"Type": "amount_v2"` on the wire while `field_type()` still returns `"number"`. Fixed by deriving the checked type from `serialize_to_map()`'s actual `Type` entry (new `wire_type()` helper) instead of `field_type()`, and reused it for the `field_type` reported in findings. This tracks whatever reaches the wallet regardless of future remaps.

File: src/visualsign/src/anchorage_render.rs (hunk 4 above)
Comment (prasanna-anchorage) [reply src/visualsign/src/anchorage_render.rs]: Confirmed — `create_preview_layout` always leaves `condensed: None`, and several `erc20.rs` call sites leave `expanded: None` too, so this is a live parser output shape today, not hypothetical. Took the second option: added `wire_has_key()` to check the actual serialized `Condensed`/`Expanded` keys rather than the in-memory `Option`, plus a new `MissingRequiredList` reason and two regression tests (missing one list, missing both). Checking the wire output rather than hardcoding "`None` = broken" means this stays correct on its own once #403 changes the serializer to always emit both keys — no merge-order dependency either way.

File: src/visualsign/examples/check_anchorage_render.rs (hunk 5 above)
Comment (prasanna-anchorage) [reply src/visualsign/examples/check_anchorage_render.rs:5]: Fixed — the deserialize now maps to an error that includes a hint about `--features diagnostics` when parsing fails, instead of surfacing the bare serde "unknown variant" error.

File: src/visualsign/src/anchorage_render.rs (hunk 4 above)
Comment (prasanna-anchorage) [reply src/visualsign/src/anchorage_render.rs]: Update after rebasing onto main: #403 merged, and its `SignablePayloadFieldPreviewLayout` serializer always emits both `Condensed` and `Expanded` (empty list when unset) — so the missing-list failure mode this thread flagged is fixed at the serialization layer, not just documented as pending. That makes `wire_has_key()`/`MissingRequiredList` unreachable: `wire_has_key` can never return false once #403 is in the tree. Removed both, and replaced the two `MissingRequiredList` tests with `preview_layout_with_unset_lists_renders_clean` / `preview_layout_with_both_lists_unset_renders_clean`, which pin that a `None` condensed/expanded (as `create_preview_layout` and several Ethereum visualizers produce) renders clean now that the wire always carries both keys. Doc matrix and spec updated to match — no missing-list case is documented anymore.

File: src/visualsign/src/anchorage_render.rs (hunk 3 above)
Comment (prasanna-anchorage) [reply src/visualsign/src/anchorage_render.rs]: Update after rebasing onto main: #393 is merged now, and CI confirmed the exact failure mode you called out — `number_field_is_unsupported` failed post-rebase because `wire_type()` correctly returns `amount_v2` for a `Number` field once #393's `serialize_to_map` remap is in the tree. Replaced that test with `number_field_renders_as_amount_v2_on_the_wire`, a regression test asserting `Number` findings are empty and `wire_type()` returns `amount_v2`. Switched the other tests' "unsupported leaf" fixture from `number` to `divider`, which isn't remapped to a renderable type, so it stays a stable unsupported example independent of future wire remaps.

File: src/visualsign/src/anchorage_render.rs [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+            ));
+            false
+        }
+        // Every other variant is a leaf: renderable iff its wire `Type` is in
+        // the set. Checked against the wire type, not `field_type()`: a
+        // variant's wire representation can be remapped (e.g. serialized
+        // under a different `Type`) without changing what `field_type()`
+        // reports, and this must track what actually reaches the wallet.
+        _ => {
+            if ANCHORAGE_RENDERABLE_LEAF_TYPES.contains(&wire_type(field).as_str()) {
```
Comment (shahan-khatchadourian-anchorage) [inline src/visualsign/src/anchorage_render.rs:182]: nit, non-blocking: `wire_type(field)` gets called twice per unsupported leaf — once here to check set membership, and again inside `finding()` (line 218) to populate the report — each call does a full `serialize_to_map()` of the field. Could pass the already-computed string into `finding()` instead of recomputing it, since this path is hit for every field in the tree.

File: docs/specs/anchorage-wallet-render-rules.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+   renderable. A single unsupported descendant flags the whole container as
+   containing unsupported nested fields. Nesting of supported leaves is unlimited in
+   depth; nesting via `preview_layout` is the supported way to express hierarchy.
+
+3. **`preview_layout` always carries both `Condensed` and `Expanded` on the
+   wire.** The `SignablePayloadFieldPreviewLayout` serializer emits an empty
+   list for whichever is unset (see #403), so the wallet's decoder always
+   receives both — there is no missing-list failure mode for the validator to
+   catch. `create_preview_layout` always leaves `Condensed` unset in memory,
+   and several Ethereum visualizers (e.g. the ERC-20 `Transfer` preview) leave
```
Comment (shahan-khatchadourian-anchorage) [inline docs/specs/anchorage-wallet-render-rules.md:72]: nit, non-blocking: this says the ERC-20 `Transfer` preview "leave[s] `Expanded` unset too" — `Transfer` actually only leaves `Condensed` unset; it's the read-only query calls (Name/Symbol/Decimals/TotalSupply) that leave both unset. Doesn't matter functionally anymore since #403 makes both cases render clean either way, but the citation is slightly off.

File: src/visualsign/examples/check_anchorage_render.rs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+use std::fs;
+
+use visualsign::SignablePayload;
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let path = std::env::args()
+        .nth(1)
+        .ok_or("usage: check_anchorage_render <payload.json>")?;
+    let json = fs::read_to_string(&path)?;
+    let payload: SignablePayload = serde_json::from_str(&json).map_err(|e| {
```
Comment (shahan-khatchadourian-anchorage) [inline src/visualsign/examples/check_anchorage_render.rs:16]: nit, non-blocking: this hint is appended to *any* parse failure, including ones unrelated to the diagnostics feature (a plain JSON syntax error, or an unknown `Type` string that isn't `diagnostic`). Could make it conditional on the error actually being an unrecognized-variant/diagnostic-shaped failure so it doesn't send someone down the wrong path for an unrelated JSON error.

**Review by shahan-khatchadourian-anchorage — APPROVED** [review]
Approving — the two substantive issues on this thread (the #393 `number`/`amount_v2` wire-type remap and the #403 missing-`Condensed`/`Expanded` fix) are both properly resolved at the root cause (serializer-level fix for the latter, which is better than just documenting it), the now-dead `MissingRequiredList`/`wire_has_key` machinery was cleanly removed, and all 9 tests + clippy pass locally.

Did a second pass for anything not already covered. Nothing blocking — three small nits below, all non-blocking.
